marda_registry/data/extractors/quantumespresso.yml

```
@@ -0,0 +1,40 @@
+---
+id: qe-tools
+name: Tools for Quantum ESPRESSO
+description: A set of tools for the Quantum ESPRESSO suite of electronic-structure simulation codes.
+subject:
+    - atomistic simulation
+    - electronic structure calculations
+    - materials modeling
+source_repository: https://github.com/aiidateam/qe-tools
+usage:
+    - method: python
+      setup: qe_tools
```

Running into one more issue with the git version, that I cannot access `qe_tools.extractors` with a simple `import qe_tools`. Works fine if you directly `import qe_tools.extractors`. Guess this can be fixed at either end, as you prefer

```suggestion
      setup: qe_tools.extractors
```

---

I'll fix it on my end

---

New error:

```
ValueError: Supported parsers: ['PW', 'CP']; given: 'QE-PW-IN'
```

---

I think maybe I'm misunderstanding how this is executed. The command is `qe_tools.extractos.extract({{ input_path }}, {{ input_type }})`. Under `supported_filetypes`, I specify `input_type` to be "PW" or "CP". "QE-PW-IN" suggests the `id` is passed instead of the type?

---

Whoops, this was my bad, sorry to bother you on a Sunday, I was missing your `template` instructions from my minimal example. This is something that should work but doesn't seem to yet, I think because the API assumes that variables defined in `template` won't override the real templated values (e.g., you could define `my_input_type` and use that in your usage string and it should work). I'll try to fix this next week

---

I'm clearly working on a Sunday 😅

---

End-to-end example works nicely now I've actually implemented the templates...

```python
$ python
Python 3.11.7 (main, Dec 13 2023, 22:06:27) [GCC 13.2.1 20230801] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from marda_extractors_api import extract
>>> extract("https://raw.githubusercontent.com/marda-alliance/metadata_extractors_registry/4e04310205ded7fe6ce7b6cb3a6b9b61895164ce/marda_registry/data/lfs/qe-pw-in/pw.in", "qe-pw-in", install=True, use_venv=True, extractor_definition={"id": "qe-tools", "usage": [{"method": "python", "setup": "qe_tools", "command": "qe_tools.extractors.extract({{ input_path }}, {{ input_type }})"}], "installation": [{"method": "pip", "requires_python": ">=3.8", "packages": ["git+https://github.com/aiidateam/qe-tools.git"]}], "supported_filetypes": [{"id": "qe-pw-in", "template": {"input_type": "PW"}}]})
Attempting to install qe-tools
Collecting git+https://github.com/aiidateam/qe-tools.git
  Cloning https://github.com/aiidateam/qe-tools.git to /tmp/pip-req-build-qxh8krw_
  Running command git clone --filter=blob:none --quiet https://github.com/aiidateam/qe-tools.git /tmp/pip-req-build-qxh8krw_
  Resolved https://github.com/aiidateam/qe-tools.git to commit 8c50b4c6e203b1ccea7dd90f384c2e365e74dd5c
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools==2.1.0) (1.26.3)
Requirement already satisfied: scipy in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools==2.1.0) (1.11.4)
Requirement already satisfied: packaging in ./marda-venvs/env-qe-tools/lib/python3.11/site-packages (from qe-tools==2.1.0) (23.2)

[notice] A new release of pip is available: 23.2.1 -> 23.3.2
[notice] To update, run: /home/mevans/repos/marda/metadata_extractors_api/marda-venvs/env-qe-tools/bin/python -m pip install --upgrade pip
/home/mevans/.pyenv/versions/3.11.7/lib/python3.11/multiprocessing/resource_tracker.py:254: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
{'structure': {'positions': [[0.005238854365041, 0.005238854365041, 0.0], [0.9697172347411749, -0.11869444788673698, -5.2917720858999996e-05], [-0.11869444788673698, 0.9697172347411749, 0.00010583544171799999]], 'species': {'names': ['O', 'H'], 'masses': [16.0, 1.0], 'pseudo_file_names': ['O.blyp-mt.UPF', 'H.blyp-vbc.UPF']}, 'cell': [[6.350126503079999, 0.0, 0.0], [0.0, 6.350126503079999, 0.0], [0.0, 0.0, 6.350126503079999]], 'atom_names': ['O', 'H', 'H']}, 'k-points': {'type': 'gamma'}}
```

---

Not sure what's all that about a memory leak, but good? 👍?

---

:shushing_face:

---

🤐